File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      is_silent? input -> "Fine. Be that way!"
      is_yelling? input -> "Whoa, chill out!"
      is_question? input -> "Sure."
      true -> "Whatever."
    end
  end

  def is_silent?(str) do
    String.strip(str) |> String.length == 0
  end

  def is_yelling?(str) do
    (has_char(str) or has_utf8(str)) and String.upcase(str) == str
  end

  def is_question?(str) do
    String.last(str) == "?"
  end

  def has_char(str) do
    String.match?(str, ~r/[a-zA-z]/)
  end

  def has_utf8(str) do
    str |> String.codepoints |> Enum.any?(fn x -> byte_size(x) == 2 end)
  end
end

File: ruby/bob/bob.rb
class Bob
  def hey remark
    builder = ResponseBuilder.new(remark)
    response = builder.build
    response.to_s
  end
end

class ResponseBuilder
  def initialize remark
    @remark = remark
  end

  def build
    case
    when Silent.is_type?(@remark)
      Silent.new
    when Yelling.is_type?(@remark)
      Yelling.new
    when Question.is_type?(@remark)
      Question.new
    else
      Other.new
    end
  end
end

class Response
  def type
    self.class.name
  end
  def self.is_type? remark; raise end
  def to_s; raise; end
end

class Silent < Response
  def self.is_type? remark
    remark.strip.length == 0
  end
  def to_s
    "Fine. Be that way!"
  end
end

class Yelling < Response
  def self.is_type? remark
    remark =~ /[a-zA-z]/ && remark.upcase == remark
  end
  def to_s
    "Whoa, chill out!"
  end
end

class Question < Response
  def self.is_type? remark
    remark[-1] == "?"
  end
  def to_s
    "Sure."
  end
end

class Other < Response
  def to_s
    "Whatever."
  end
end
